consensus/istanbul/backend/announce.go

```diff
@@ -700,3 +719,180 @@ func (sb *Backend) checkPeersAnnounceVersions() {
 		}
 	}
 }
+
+type versionedEnode struct {
+	Node    string
+	Version uint
+}
+
+// ==============================================
+//
+// define the functions that needs to be provided for rlp Encoder/Decoder.
+
+// EncodeRLP serializes ve into the Ethereum RLP format.
+func (ve *versionedEnode) EncodeRLP(w io.Writer) error {
+	return rlp.Encode(w, []interface{}{ve.Node, ve.Version})
+}
+
+// DecodeRLP implements rlp.Decoder, and load the ve fields from a RLP stream.
+func (ve *versionedEnode) DecodeRLP(s *rlp.Stream) error {
+	var msg struct {
+		Node    string
+		Version uint
+	}
+
+	if err := s.Decode(&msg); err != nil {
+		return err
+	}
+	ve.Node, ve.Version = msg.Node, msg.Version
+	return nil
+}
+
+// getSelfVersionedEnodeMsg gets the most recent self versioned enode message to send
+// and generates a new one with `version` if one does not exist.
+// New versioned enode messages are not always generated to ensure the version
+// of a versioned enode message is not greater than a recently gossiped announce
+// version.
+// May be nil if this is a proxy that does not have a versioned enode message
+// from its proxy.
+func (sb *Backend) getSelfVersionedEnodeMsg(version uint) (*istanbul.Message, error) {
```

Comment:
I find this interface somewhat misleading because it is titled as a `get` function, but it can end up modifying the Backend state. It could be considered a cache, but it is not the only place this is set, and the result of the calculation is dependent on time. Namely if this node have not yet gossiped out any announce messages, this function will break the invariant that is stated above that the version should not be newer than the gossiped announce. I also thought when first read the function that the `version` arg is used in retrieval (i.e. that the returned message version would be the same as the version arg) but that is not the case.

Reply:
Yeah, I agree this is pretty confusing. For the time being I renamed `getSelfVersionedEnodeMsg` -> `retrieveSelfVersionedEnodeMsg` to make it feel like less of a getter, and renamed `version` to `fallbackVersion`. I also clarified the invariant you mentioned
I couldn't think of a good way to refactor, but I could spend more time if you don't think those small changes are enough

Reply:
Just refactored in a better way and removed the need for the fallbackVersion